daemon/emer-circular-file.c

```diff
@@ -0,0 +1,825 @@
+ * successful call to emer_circular_file_remove invalidates any outstanding
+ * tokens. If no elements were read but the read succeeded, then elems is set to
+ * NULL. Returns TRUE on success and FALSE on error.
+ */
+gboolean
+emer_circular_file_read (EmerCircularFile *self,
+                         GBytes         ***elems,
+                         gsize             data_bytes_to_read,
+                         gsize            *num_elems,
+                         guint64          *token,
+                         GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  if (priv->size == 0)
+    {
+      *elems = NULL;
+      *num_elems = 0;
+      *token = 0;
+      return TRUE;
+    }
+
+  GFileInputStream *file_input_stream =
+    g_file_read (priv->data_file, NULL /* GCancellable */, error);
+  if (file_input_stream == NULL)
+    return FALSE;
+
+  gboolean seek_succeeded =
+    g_seekable_seek (G_SEEKABLE (file_input_stream), priv->head, G_SEEK_SET,
+                     NULL /* GCancellable */, error);
+  if (!seek_succeeded)
+    {
+      g_object_unref (file_input_stream);
+      return FALSE;
+    }
+
+  GPtrArray *elem_array =
+    g_ptr_array_new_with_free_func ((GDestroyNotify) g_bytes_unref);
+
+  guint64 curr_data_bytes = 0;
+  guint64 curr_disk_bytes = 0;
+  GInputStream *input_stream = G_INPUT_STREAM (file_input_stream);
+  while (curr_disk_bytes < priv->size)
+    {
+      guint64 elem_size;
+      if (!read_elem_size (self, input_stream, &elem_size, error))
+        goto handle_failed_read;
+
+      guint64 next_data_bytes = curr_data_bytes + elem_size;
+      if (next_data_bytes > data_bytes_to_read)
+        break;
+
+      gboolean append_succeeded =
+        append_elem_to_array (self, input_stream, elem_array, elem_size, error);
+      if (!append_succeeded)
+        goto handle_failed_read;
+
+      curr_data_bytes = next_data_bytes;
+      curr_disk_bytes += sizeof (elem_size) + elem_size;
+    }
+
+  g_object_unref (file_input_stream);
+
+  *num_elems = elem_array->len;
+  *elems = (GBytes **) g_ptr_array_free (elem_array, FALSE);
+
+  *token = curr_disk_bytes;
+
+  return TRUE;
+
+handle_failed_read:
+  g_object_unref (file_input_stream);
+  g_ptr_array_unref (elem_array);
+  return FALSE;
+}
+
+/* Removes the elements that were read in the call to emer_circular_file_read
+ * that produced the given token. Tokens may not be reused, and any successful
+ * call to emer_circular_file_remove invalidates any outstanding tokens. A token
+ * of 0 indicates that no elements should be removed. Returns TRUE on success
+ * and FALSE on error.
+ */
+gboolean
+emer_circular_file_remove (EmerCircularFile *self,
+                           guint64           token,
+                           GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  g_return_val_if_fail (token <= priv->size, FALSE);
```

Comment:
Either always consistently check in all public methods for this or never. I don't think you should, since this is not a library.